### PR TITLE
Improvements to single AZ packing

### PR DESCRIPTION
**Pull request by pisarenko-net**
This PR makes changes to the single AZ packing function.

Most importantly, the zones are now considered in the expected order. Previously, this was left to chance as the order of the keys in the map is unpredictable. Furthermore, all AZs are tried and the one with the highest efficiency is chosen. This is because choosing the first AZ may yield a less efficient outcome.

There are changes to the API. The results are returned in a struct. Packing efficiency is included. This is to enable upstream metrics.

Disclaimer: this is my first production use of golang so please watch out.

### CODE REVIEW

**Review by Chrisbattarbee — CHANGES_REQUESTED** [review]
Thanks for submitting this PR, I think it has a lot of good ideas we can use. Especially emitting efficiency for binpacks however I think we should be more targeted with smaller PRs.

I think the first PR should be simply to improve single AZ packing by considering the packed utilisation of hosts we would use. Then in another PR we can add the overall efficiency metrics and potentially restructure the PackingResult

File: pkg/binpack/pack_tightly.go [hunk 1]
```diff
@@ -50,6 +50,7 @@ func tightlyPackExecutors(
 			reservedResources[n].Add(executorResources)
 			nodeSchedulingMetadata, ok := nodesSchedulingMetadata[n]
 			if !ok || reservedResources[n].GreaterThan(nodeSchedulingMetadata.AvailableResources) {
+				reservedResources[n].Sub(executorResources)
```
[the remark this answers is not in the document]
Comment (pisarenko-net) [reply pkg/binpack/pack_tightly.go:53]: Ditto `distribute_evenly.go`.

File: pkg/binpack/single_az_pack_tightly.go [hunk 2]
```diff
@@ -29,33 +29,53 @@ var SingleAZTightlyPack = SparkBinPackFunction(func(
...
 	nodeNamesByZone := make(map[string][]string)
 	for _, nodeName := range nodeNames {
 		nodesSchedulingMetadata, ok := nodesSchedulingMetadata[nodeName]
 		if !ok {
 			continue
 		}
 		zoneLabel := nodesSchedulingMetadata.ZoneLabel
+		if _, ok := nodeNamesByZone[zoneLabel]; !ok {
+			zonesInOrder = append(zonesInOrder, zoneLabel)
+		}
```
[the remark this answers is not in the document]
Comment (pisarenko-net) [reply pkg/binpack/single_az_pack_tightly.go:77]: That's right!

File: pkg/binpack/efficiency.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	Memory   float64
+	GPU      float64
+}
+
+// LessThan compares two packing efficiencies. For a single packing we take the highest of the
+// resources' efficiency. For example, when CPU is at 0.81 and Memory is at 0.54 the avg efficiency
+// is 0.81. One packing efficiency is deemed less efficient when its avg efficiency is lower than
+// the other's packing efficiency.
+func (p *PackingEfficiency) LessThan(o PackingEfficiency) bool {
+	// TODO: GPU is explicitly excluded for now but worthwhile to reconsider in future
```
[the remark this answers is not in the document]
Comment (pisarenko-net) [reply pkg/binpack/efficiency.go:37]: I propose to represent a missing GPU with a different value (0 or -1), so that it does not affect this calculation and we don't ever need to come back to it.

Will send an update once I consider different options.

File: pkg/binpack/efficiency.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+)
+
+// PackingEfficiency represents result packing efficiency per resource type. Computed as the total
+// resources used divided by total capacity.
+type PackingEfficiency struct {
+	NodeName string
+	CPU      float64
+	Memory   float64
+	GPU      float64
+}
```
[the remark this answers is not in the document]
Comment (pisarenko-net) [reply pkg/binpack/efficiency.go]: Agreed. Resolved.

File: pkg/binpack/efficiency.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+	// TODO: GPU is explicitly excluded for now but worthwhile to reconsider in future
+	pMaxEfficiency := math.Max(p.CPU, p.Memory)
+	oMaxEfficiency := math.Max(o.CPU, o.Memory)
+	return pMaxEfficiency < oMaxEfficiency
+}
+
+// ComputePackingEfficiencies calculates utilization for all provided nodes, given the new reservation.
+func ComputePackingEfficiencies(
+	nodeGroupSchedulingMetadata resources.NodeGroupSchedulingMetadata,
+	reservedResources resources.NodeGroupResources) (PackingEfficiency, []*PackingEfficiency) {
```
[the remark this answers is not in the document]
Comment (pisarenko-net) [reply pkg/binpack/efficiency.go]: Agreed fully. Resolved.

File: pkg/binpack/binpack.go [hunk 6]
```diff
@@ -20,13 +20,31 @@ import (
...
+	PackingEfficiencies  []*PackingEfficiency
+	HasCapacity          bool
+}
+
+func emptyPackingResult() *PackingResult {
+	return &PackingResult{
+		DriverNode:    "",
+		ExecutorNodes: nil,
+		HasCapacity:   false,
+	}
```
[the remark this answers is not in the document]
Comment (pisarenko-net) [reply pkg/binpack/binpack.go]: Done.

File: pkg/binpack/binpack.go [hunk 7]
```diff
@@ -44,19 +62,26 @@ func SparkBinPack(
 	executorCount int,
 	driverNodePriorityOrder, executorNodePriorityOrder []string,
 	nodesSchedulingMetadata resources.NodeGroupSchedulingMetadata,
-	distributeExecutors GenericBinPackFunction) (driverNode string, executorNodes []string, hasCapacity bool) {
-	for _, name := range driverNodePriorityOrder {
-		nodeSchedulingMetadata, ok := nodesSchedulingMetadata[name]
+	distributeExecutors GenericBinPackFunction) *PackingResult {
+	for _, driverNode := range driverNodePriorityOrder {
```
[the remark this answers is not in the document]
Comment (pisarenko-net) [reply pkg/binpack/binpack.go]: Good catch.

File: pkg/binpack/single_az_pack_tightly.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pisarenko-net) [reply pkg/binpack/single_az_pack_tightly.go:77]: Is something concerning you here?

Without deterministic order we're basically screwing up the prior node ordering, at least when it comes to AZ order. The order of nodes within AZ was not previously affected.

**Comment by pisarenko-net** [comment]
> Thanks for submitting this PR, I think it has a lot of good ideas we can use. Especially emitting efficiency for binpacks however I think we should be more targeted with smaller PRs.
> 
> I think the first PR should be simply to improve single AZ packing by considering the packed utilisation of hosts we would use. Then in another PR we can add the overall efficiency metrics and potentially restructure the PackingResult

Thank you for the feedback. The reason for introducing a new PackingEfficiency struct is to simplify packing assessment. Without the new primitives every packing function would need to compute packing efficiencies by themselves. For this reason the order of splitting up this PR is to first release the refactor, then the changes to packing function(s).

File: pkg/binpack/single_az_pack_tightly.go [hunk 8]
```diff
@@ -29,33 +29,53 @@ var SingleAZTightlyPack = SparkBinPackFunction(func(
...
+		return emptyPackingResult()
+	}
+
+	// choose the most efficient packing across AZs
+	bestResult := packingResults[0]
+	for _, result := range packingResults {
+		if bestResult.AvgPackingEfficiency.LessThan(result.AvgPackingEfficiency) {
+			bestResult = result
+		}
+	}
```
[the remark this answers is not in the document]
Comment (pisarenko-net) [reply pkg/binpack/single_az_pack_tightly.go]: Makes sense. I misunderstood our previous conversation. Resolved.

File: pkg/binpack/distribute_evenly.go [hunk 9]
```diff
@@ -60,6 +60,7 @@ func distributeExecutorsEvenly(
 			if !ok || reservedResources[n].GreaterThan(nodeSchedulingMetadata.AvailableResources) {
 				// can not allocate a resource to this node
 				delete(availableNodes, n)
+				reservedResources[n].Sub(executorResources)
```
[the remark this answers is not in the document]
Comment (Chrisbattarbee) [reply pkg/binpack/distribute_evenly.go:63]: Ah gotcha

File: pkg/binpack/distribute_evenly_test.go [hunk 10]
```diff
@@ -30,10 +31,23 @@ func createResources(cpu, memory, nvidiaGPU int64) *resources.Resources {
...
+
+func createSchedulingMetadataWithTotals(
+	availableCPU, cpuTotal, availableMemory, memoryTotal, availableNvidiaGPU, nvidiaGPUTotal int64,
+	zoneLabel string) *resources.NodeSchedulingMetadata {
+
+	return &resources.NodeSchedulingMetadata{
+		AvailableResources:   createResources(availableCPU, availableMemory, availableNvidiaGPU),
+		SchedulableResources: createResources(cpuTotal, memoryTotal, nvidiaGPUTotal),
+		ZoneLabel:            zoneLabel,
 	}
```
Comment (Chrisbattarbee) [inline pkg/binpack/distribute_evenly_test.go:51]: Lets move this closer to where it's actually used

File: pkg/binpack/efficiency.go [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+
+package binpack
+
+import (
+	"math"
+
+	"github.com/palantir/k8s-spark-scheduler-lib/pkg/resources"
+)
+
+// AvgPackingEfficiency represents result packing efficiency per resource type for the entire cluster.
```
Comment (Chrisbattarbee) [inline pkg/binpack/efficiency.go:23]: This isnt the entire cluster right?

File: pkg/binpack/efficiency.go [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+// PackingEfficiency represents result packing efficiency per resource type for one node. Computed
+// as the total resources used divided by total capacity.
+type PackingEfficiency struct {
+	NodeName string
+	CPU      float64
+	Memory   float64
+	GPU      float64
+}
+
+// Max returns the highest packing efficiency of CPU and Memory.
```
Comment (Chrisbattarbee) [inline pkg/binpack/efficiency.go:60]: Looks like this is also GPU now

File: pkg/binpack/efficiency.go [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+func (p *PackingEfficiency) Max() float64 {
+	return math.Max(p.GPU, math.Max(p.CPU, p.Memory))
+}
+
+// ComputePackingEfficiencies calculates utilization for all provided nodes, given the new reservation.
+func ComputePackingEfficiencies(
+	nodeGroupSchedulingMetadata resources.NodeGroupSchedulingMetadata,
+	reservedResources resources.NodeGroupResources) map[string]*PackingEfficiency {
```
Comment (Chrisbattarbee) [inline pkg/binpack/efficiency.go:68]: I think you'll want to pass pointers to these?

File: pkg/binpack/efficiency.go [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+
+func normalizeResource(resourceValue int64) int64 {
+	if resourceValue == 0 {
+		return 1
+	}
+	return resourceValue
+}
+
+// ComputeAvgPackingEfficiency calculate average packing efficiency for an entire cluster, given
+// packing efficiencies for individual nodes.
```
Comment (Chrisbattarbee) [inline pkg/binpack/efficiency.go:113]: This is just for a group of nodes, not the cluster right?

File: pkg/binpack/single_az_pack_tightly_test.go [hunk 15]
```diff
@@ -88,19 +88,37 @@ func TestSingleAZTightlyPack(t *testing.T) {
...
+			"n2_z1": createSchedulingMetadataWithTotals(0, 10, 0, 0, 0, 0, "z1"),
+			"n3_z1": createSchedulingMetadataWithTotals(10, 10, 0, 0, 0, 0, "z1"),
+			"n1_z2": createSchedulingMetadataWithTotals(9, 10, 0, 0, 0, 0, "z2"),
+			"n2_z2": createSchedulingMetadataWithTotals(10, 10, 0, 0, 0, 0, "z2"),
+			"n3_z2": createSchedulingMetadataWithTotals(10, 10, 0, 0, 0, 0, "z2"),
+		}),
+		nodePriorityOrder:      []string{"n1_z1", "n2_z1", "n3_z1", "n1_z2", "n2_z2", "n3_z2"},
+		expectedDriverNode:     "n1_z2",
+		willFit:                true,
+		expectedExecutorCounts: map[string]int{},
```
Comment (Chrisbattarbee) [inline pkg/binpack/single_az_pack_tightly_test.go:107]: Can we have a few more larger examples with executors to make sure we're covering all our bases here

**Review by Chrisbattarbee — CHANGES_REQUESTED** [review]
Okay looking nice just a few comments!

File: pkg/binpack/efficiency.go (hunk 12 above)
Comment (pisarenko-net) [reply pkg/binpack/efficiency.go:60]: Well caught. Fixed.

File: pkg/binpack/efficiency.go (hunk 14 above)
Comment (pisarenko-net) [reply pkg/binpack/efficiency.go:113]: Absolutely. Fixed.

File: pkg/binpack/efficiency.go (hunk 13 above)
Comment (pisarenko-net) [reply pkg/binpack/efficiency.go:68]: Do you mean `reservedResources`? The type `resources.NodeGroupResources` is defined as `type NodeGroupResources map[string]*Resources`.

`nodeGroupSchedulingMetadata` is a struct. The resource fields are pointers though.

Asking for a clarification.

File: pkg/binpack/distribute_evenly_test.go (hunk 10 above)
Comment (pisarenko-net) [reply pkg/binpack/distribute_evenly_test.go:51]: Ack. Moved this to resources where it should belong.

File: pkg/binpack/single_az_pack_tightly_test.go (hunk 15 above)
Comment (pisarenko-net) [reply pkg/binpack/single_az_pack_tightly_test.go:107]: I agree. Adding a couple of test cases.

Comment (pisarenko-net) [reply pkg/binpack/single_az_pack_tightly_test.go:107]: Added a carefully chosen corner case test. AZ1 yields better average packing efficiency across AZ and across all nodes. AZ2 yields better average packing efficiency over 2 nodes in AZ2 but otherwise worse.

Comment (Chrisbattarbee) [reply pkg/binpack/single_az_pack_tightly_test.go:107]: Nice

**Review by Chrisbattarbee — APPROVED** [review]
